os/se/ameba/rtl_security_api_wrapper.c

```diff
@@ -0,0 +1,392 @@
+			ret = RTL_HAL_NOT_SUPPORTED;
+		}
+	}
+	device_mutex_unlock(RT_DEV_LOCK_CRYPTO);
+
+	return ret;
+}
+
+uint32_t rtl_cryptoAES_ctr_wrapper(uint8_t* key, uint32_t keylen, unsigned char* message, uint32_t msglen, unsigned char* pResult, uint8_t mode)
+{
+	uint32_t ret = RTL_HAL_SUCCESS;
+	unsigned char iv[16];
+
+	memcpy(iv, aes_iv, sizeof(aes_iv));
+	device_mutex_lock(RT_DEV_LOCK_CRYPTO);
+	ret = rtl_crypto_aes_ctr_init(key, keylen);
+
+	if (ret == RTL_HAL_SUCCESS) {
+		if (mode == MBEDTLS_ENCRYPT) {
+			//Encryption operation
+			ret = rtl_crypto_aes_ctr_encrypt(message, msglen, iv, sizeof(iv), pResult);
+		} else if (mode == MBEDTLS_DECRYPT) {
+			//Decryption operation
+			ret = rtl_crypto_aes_ctr_decrypt(message, msglen, iv, sizeof(iv), pResult);
+		} else {
+			ret = RTL_HAL_NOT_SUPPORTED;
+		}
+	}
+	device_mutex_unlock(RT_DEV_LOCK_CRYPTO);
+
+	return ret;
+}
+
+/**
+ * Secure Storage
+ */
+uint32_t rtl_write_storage_wrapper(uint32_t ss_idx, hal_data *data)
+{
+	uint32_t	ret;
+	uint8_t		tmpbuf[64];
+	uint8_t		buf_length[4];
+	uint8_t		*key = (uint8_t*)(((uint32_t) tmpbuf + 31) & (~31));	// align to 32byte
+	uint32_t	data_len = (data->data_len + 15) / 16 * 16;				// AES blocks
+	flash_t		flash;
+	const uint8_t aes_gcm_iv[12];
+	const uint8_t aes_gcm_aad[20];
+
+	uint8_t	*buf = (uint8_t *) malloc(data_len + TAG_LENGTH);
+	uint8_t	*buf_flash = (uint8_t *) malloc(data_len + TAG_LENGTH + 4);
+	uint8_t	*data_plain = (uint8_t *) malloc(data_len);
+	uint8_t	*tag = buf + data_len;
+
+	memcpy(aes_gcm_iv, aes_test_iv_gcm, sizeof(aes_test_iv_gcm));
+	memcpy(aes_gcm_aad, aes_test_aad_gcm, sizeof(aes_test_aad_gcm));
+
+	if (buf && buf_flash && data_plain) {
+		memset(buf, 0, data_len + TAG_LENGTH);
+		memset(buf_flash, 0, data_len + TAG_LENGTH + 4);
+		memset(data_plain, 0, data_len);
+		memcpy(data_plain, data->data, data->data_len);
+
+		if (!crypto_inited) {
+			device_mutex_lock(RT_DEV_LOCK_CRYPTO);
+			ret = rtl_cryptoEngine_init();
+			device_mutex_unlock(RT_DEV_LOCK_CRYPTO);
+			if (ret == RTL_HAL_SUCCESS) {
+				crypto_inited = 1;
+			} else {
+				sedbg("\n\r ERROR: rtl_cryptoEngine_init \n\r");
+				goto exit;
+			}
+		}
+		memset(key, 0, sizeof(key));
+		memcpy(key, samsung_key, SAMSUNG_KEY_SIZE);
+
+		device_mutex_lock(RT_DEV_LOCK_CRYPTO);
+		ret = rtl_crypto_aes_gcm_init(key, 32);
+		device_mutex_unlock(RT_DEV_LOCK_CRYPTO);
+		if (ret != 0) {
+			sedbg("\n\r ERROR: crypto_aes_gcm_init \n\r");
+			goto exit;
+		}
+		device_mutex_lock(RT_DEV_LOCK_CRYPTO);
+		ret = rtl_crypto_aes_gcm_encrypt(data_plain, data_len, aes_gcm_iv, aes_gcm_aad, sizeof(aes_gcm_aad), buf, tag);
+		device_mutex_unlock(RT_DEV_LOCK_CRYPTO);
+
+		if (ret != 0) {
+			sedbg("\n\r ERROR: crypto_aes_gcm_encrypt \n\r");
+			goto exit;
+		}
+		itoa(data_len, buf_length, 10);
+		memcpy(buf_flash, buf_length, 4);
+		memcpy(buf_flash + 4, buf, data_len + TAG_LENGTH);
+
+		device_mutex_lock(RT_DEV_LOCK_FLASH);
+		flash_erase_sector(&flash, SEC_STORE_OFFSET + ss_idx * 4096);
```

Comment:
Please check if exit need sto be calle din case of failure of these API calls as well

Reply:
@thapav these API's do not provide a return value hence they arent checkked.